Infrastructure/Services/ProcessEmailsService.cs

```diff
@@ -0,0 +1,486 @@
+                Only if the status has changed do the following:
+                    a) Update status and LastStatusChangedDate = now
+                    b) If new status is NotifyDelivered see if any chained emails (in EmailTriggerActivation and status WaitingToBeTriggered).  If so update these ProcessFromDate = now
+                    c) If the new status is failure (If RetrySendInSeconds != null) then update the ProcessFromDate = now.AddSeconds(RetrySendInSeconds)
+         */
+
+        var pollDateTime = DateTime.UtcNow;
+
+        var emailsToPollForStatusUpdate = await _unitOfWork.EmailLogRepository
+            .GetAllAsync(x => (x.EmailStatus.PollForStatusUpdateIfSent &&
+                                x.LastEmailSendAttemptDate != null &&
+                                x.LastEmailSendAttemptDate < pollDateTime &&
+                                x.FinishProcessingDate > pollDateTime),
+            "EmailStatus,EmailNotifyResponseLog,EmailsActivatedByThisEmail,EmailsActivatedByThisEmail.ActivateEmailLog,EmailsActivatedByThisEmail.ActivateEmailLog.EmailStatus",
+            true);
+
+        if (emailsToPollForStatusUpdate != null && emailsToPollForStatusUpdate.Any())
+        {
+            var emailStatuses = _unitOfWork.EmailStatusRepository.GetAll();
+
+            foreach (var emailLog in emailsToPollForStatusUpdate)
+            {
+                try
+                {
+                    if (emailLog == null || emailLog.EmailNotifyResponseLog == null || emailLog.EmailStatus == null)
+                    {
+                        throw new InvalidDataException("The emailsToPollForStatusUpdate has an unexpected empty data");
+                    }
+
+                    processedEmailsModel.EmailsCheckStatus++;
+
+                    var currentStatus = emailLog.EmailStatus.Status.GetEnumFromDisplayName<EmailStatus>();
+                    var newStatus = await _notificationsClientService.GetEmailStatus(emailLog!.EmailNotifyResponseLog!.NotifyId!);
+
+                    if (currentStatus != newStatus)
+                    {
+                        ProcessUpdatedEmailStatus(emailLog, pollDateTime, currentStatus, newStatus, emailStatuses);
+
+                        _unitOfWork.EmailLogRepository.Update(emailLog);
+
+                        //Save per record rather than all in one go, so that it will process other emails and not fail for all
+                        await _unitOfWork.Complete();
+
+                        processedEmailsModel.EmailsUpdatedStatus++;
+                    }
+                }
+                catch (Exception ex)
+                {
+                    //We don't want to throw the error here, we log and continue processing other emails
+                    _logger.LogError(ex, "Unexpected error in emailsToPollForStatusUpdate for EmailLog.ClientReferenceNumber: {ClientReferenceNumber}, EmailLog.Id: {EmailLogId}", emailLog.ClientReferenceNumber, emailLog.Id);
```

Comment:
I am seeing the following error on my IDE: `Dereference of a possibly null reference` for the `emailLog.ClientReferenceNumber`. I suggest using the following `emailLog!.ClientReferenceNumber`.

Reply:
Will add this in to NTP-1218 which is in dev at the moment, since this has been fully tested.